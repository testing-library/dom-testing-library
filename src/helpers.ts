// Constant node.nodeType for text nodes, see:
// https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType#Node_type_constants
const TEXT_NODE = 3

function jestFakeTimersAreEnabled() {
  /* istanbul ignore else */
  // eslint-disable-next-line
  if (typeof jest !== 'undefined' && jest !== null) {
    return (
      // legacy timers
      (setTimeout as any)._isMockFunction === true ||
      // modern timers
      Object.prototype.hasOwnProperty.call(setTimeout, 'clock')
    )
  }
  // istanbul ignore next
  return false
}

function getDocument() {
  /* istanbul ignore if */
  if (typeof window === 'undefined') {
    throw new Error('Could not find default container')
  }
  return window.document
}
function getWindowFromNode(node: any) {
  if (node.defaultView) {
    // node is document
    return node.defaultView
  } else if (node.ownerDocument && node.ownerDocument.defaultView) {
    // node is a DOM node
    return node.ownerDocument.defaultView
  } else if (node.window) {
    // node is window
    return node.window
  } else if (node.ownerDocument && node.ownerDocument.defaultView === null) {
    throw new Error(
      `It looks like the window object is not available for the provided node.`,
    )
  } else if (node.then instanceof Function) {
    throw new Error(
      `It looks like you passed a Promise object instead of a DOM node. Did you do something like \`fireEvent.click(screen.findBy...\` when you meant to use a \`getBy\` query \`fireEvent.click(screen.getBy...\`, or await the findBy query \`fireEvent.click(await screen.findBy...\`?`,
    )
  } else if (Array.isArray(node)) {
    throw new Error(
      `It looks like you passed an Array instead of a DOM node. Did you do something like \`fireEvent.click(screen.getAllBy...\` when you meant to use a \`getBy\` query \`fireEvent.click(screen.getBy...\`?`,
    )
  } else if (
    typeof node.debug === 'function' &&
    typeof node.logTestingPlaygroundURL === 'function'
  ) {
    throw new Error(
      `It looks like you passed a \`screen\` object. Did you do something like \`fireEvent.click(screen, ...\` when you meant to use a query, e.g. \`fireEvent.click(screen.getBy..., \`?`,
    )
  } else {
    // The user passed something unusual to a calling function
    throw new Error(
      `The given node is not an Element, the node type is: ${typeof node}.`,
    )
  }
}

function checkContainerType(container: unknown) {
  if (
    !container ||
    !(typeof (container as any).querySelector === 'function') ||
    !(typeof (container as any).querySelectorAll === 'function')
  ) {
    throw new TypeError(
      `Expected container to be an Element, a Document or a DocumentFragment but got ${getTypeName(
        container,
      )}.`,
    )
  }

  function getTypeName(object: unknown) {
    if (typeof object === 'object') {
      return object === null ? 'null' : object.constructor.name
    }
    return typeof object
  }
}

export {
  getWindowFromNode,
  getDocument,
  checkContainerType,
  jestFakeTimersAreEnabled,
  TEXT_NODE,
}
