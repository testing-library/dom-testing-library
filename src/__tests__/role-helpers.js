import {
  getRoles,
  logRoles,
  getImplicitAriaRoles,
  isInaccessible,
} from '../role-helpers'
import {render} from './helpers/test-utils'

beforeEach(() => {
  jest.spyOn(console, 'log').mockImplementation(() => {})
})

afterEach(() => {
  console.log.mockRestore()
})

function setup() {
  const {getByTestId} = render(`
<header data-testid="a-header">Banner header</header>
<section aria-label="a region" data-testid='named-section'>
  <a href="http://whatever.com" data-testid="a-link">link</a>
  <a>invalid link</a>

  <nav data-testid='a-nav' />
  
  <h1 data-testid='a-h1'>Main Heading</h1>
  <h2 data-testid='a-h2'>Sub Heading</h2>
  <h3 data-testid='a-h3'>Tertiary Heading</h3>

  <article data-testid='a-article'>
    <!-- menuitem is currently deprecated, but is the only 
         tag currently that aria-query returns multiple roles for
         (roles: command, menuitem).
         It's used here in case a future tag also has multiple 
         roles -->
    <menuitem data-testid='a-menuitem-1'>1</menuitem> 
    <menuitem data-testid='a-menuitem-2'>2</menuitem>

    <ul data-testid='a-list'>
      <li data-testid='a-list-item-1'>Item 1</li>
      <li data-testid='a-list-item-2'>Item 2</li>
    </ul>

    <table data-testid='a-table'>
      <tbody data-testid='a-tbody'>
        <tr data-testid='a-row'>
          <td data-testid='a-cell-1'>Cell 1</td>
          <td data-testid='a-cell-2'>Cell 2</td>
          <td data-testid='a-cell-3'>Cell 3</td>
        </tr>
      </tbody>
    </table>

    <form aria-label="a form" data-testid='named-form'>
      <input type='radio' data-testid='a-radio-1' />
      <input type='radio' data-testid='a-radio-2' />
      <input type='text' data-testid='a-input-1' />
      <input type='text' data-testid='a-input-2' />
      <textarea data-testid='a-textarea'></textarea>
    </form>

    <ul data-testid='b-list'>
      <li data-testid='b-list-item-1'>Item 1</li>
      <li data-testid='b-list-item-2'>Item 2</li>
    </ul>

    <form data-testid="a-form" />
    <section data-testid="a-section" />
   </article>
   <dl>
    <dt data-testid="a-dt">Term</dt>
    <dd data-testid="a-dd">Definition</dd>
   </dl>
</section>
  `)

  return {
    unnamedSection: getByTestId('a-section'),
    namedSection: getByTestId('named-section'),
    anchor: getByTestId('a-link'),
    h1: getByTestId('a-h1'),
    h2: getByTestId('a-h2'),
    h3: getByTestId('a-h3'),
    nav: getByTestId('a-nav'),
    article: getByTestId('a-article'),
    menuItem: getByTestId('a-menuitem-1'),
    menuItem2: getByTestId('a-menuitem-2'),
    aUl: getByTestId('a-list'),
    aLi1: getByTestId('a-list-item-1'),
    aLi2: getByTestId('a-list-item-2'),
    bUl: getByTestId('b-list'),
    bLi1: getByTestId('b-list-item-1'),
    bLi2: getByTestId('b-list-item-2'),
    table: getByTestId('a-table'),
    tbody: getByTestId('a-tbody'),
    tr: getByTestId('a-row'),
    td1: getByTestId('a-cell-1'),
    td2: getByTestId('a-cell-2'),
    td3: getByTestId('a-cell-3'),
    unnamedForm: getByTestId('a-form'),
    namedForm: getByTestId('named-form'),
    radio: getByTestId('a-radio-1'),
    radio2: getByTestId('a-radio-2'),
    input: getByTestId('a-input-1'),
    input2: getByTestId('a-input-2'),
    textarea: getByTestId('a-textarea'),
    dt: getByTestId('a-dt'),
    dd: getByTestId('a-dd'),
    header: getByTestId('a-header'),
  }
}

test('getRoles returns expected roles for various dom nodes', () => {
  const {
    anchor,
    h1,
    h2,
    h3,
    nav,
    article,
    menuItem,
    menuItem2,
    aUl,
    aLi1,
    aLi2,
    bUl,
    bLi1,
    bLi2,
    table,
    tbody,
    tr,
    td1,
    td2,
    td3,
    radio,
    radio2,
    input,
    input2,
    textarea,
    namedSection,
    namedForm,
    dd,
    dt,
    header,
  } = setup()

  expect(getRoles(namedSection)).toEqual({
    link: [anchor],
    heading: [h1, h2, h3],
    navigation: [nav],
    radio: [radio, radio2],
    article: [article],
    list: [aUl, bUl],
    listitem: [aLi1, aLi2, bLi1, bLi2],
    table: [table],
    row: [tr],
    cell: [td1, td2, td3],
    textbox: [input, input2, textarea],
    rowgroup: [tbody],
    command: [menuItem, menuItem2],
    menuitem: [menuItem, menuItem2],
    form: [namedForm],
    region: [namedSection],
    term: [dt],
    definition: [dd],
  })
  expect(getRoles(header)).toEqual({
    banner: [header],
  })
})

test('logRoles calls console.log with output from prettyRoles', () => {
  const {namedSection} = setup()
  logRoles(namedSection)
  expect(console.log).toHaveBeenCalledTimes(1)
  expect(console.log.mock.calls[0][0]).toMatchSnapshot()
})

test('getImplicitAriaRoles returns expected roles for various dom nodes', () => {
  const {namedSection, h1, unnamedForm, radio, input} = setup()

  expect(getImplicitAriaRoles(namedSection)).toEqual(['region'])
  expect(getImplicitAriaRoles(h1)).toEqual(['heading'])
  expect(getImplicitAriaRoles(unnamedForm)).toEqual([])
  expect(getImplicitAriaRoles(radio)).toEqual(['radio'])
  expect(getImplicitAriaRoles(input)).toEqual(['textbox'])
})

test.each([
  ['<div />', false],
  ['<div aria-hidden="false" />', false],
  ['<div style="visibility: visible" />', false],
  ['<div hidden />', true],
  ['<div style="display: none;"/>', true],
  ['<div style="visibility: hidden;"/>', true],
  ['<div aria-hidden="true" />', true],
])('shouldExcludeFromA11yTree for %s returns %p', (html, expected) => {
  const {container} = render(html)
  container.firstChild.appendChild(document.createElement('button'))

  expect(isInaccessible(container.querySelector('button'))).toBe(expected)
})
