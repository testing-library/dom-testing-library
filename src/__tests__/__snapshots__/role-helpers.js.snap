// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logRoles calls console.log with output from prettyRoles 1`] = `
region:

Name "a region":
<section
  aria-label="a region"
  data-testid="named-section"
/>

--------------------------------------------------
link:

Name "link":
<a
  data-testid="a-link"
  href="http://whatever.com"
/>

--------------------------------------------------
navigation:

Name "":
<nav
  data-testid="a-nav"
/>

--------------------------------------------------
heading:

Name "Main Heading":
<h1
  data-testid="a-h1"
/>

Name "Sub Heading":
<h2
  data-testid="a-h2"
/>

Name "Tertiary Heading":
<h3
  data-testid="a-h3"
/>

--------------------------------------------------
article:

Name "":
<article
  data-testid="a-article"
/>

--------------------------------------------------
command:

Name "":
<menuitem
  data-testid="a-menuitem-1"
/>

Name "":
<menuitem
  data-testid="a-menuitem-2"
/>

--------------------------------------------------
menuitem:

Name "":
<menuitem
  data-testid="a-menuitem-1"
/>

Name "":
<menuitem
  data-testid="a-menuitem-2"
/>

--------------------------------------------------
list:

Name "":
<ul
  data-testid="a-list"
/>

Name "":
<ul
  data-testid="b-list"
/>

--------------------------------------------------
listitem:

Name "":
<li
  data-testid="a-list-item-1"
/>

Name "":
<li
  data-testid="a-list-item-2"
/>

Name "":
<li
  data-testid="b-list-item-1"
/>

Name "":
<li
  data-testid="b-list-item-2"
/>

--------------------------------------------------
table:

Name "":
<table
  data-testid="a-table"
/>

--------------------------------------------------
rowgroup:

Name "":
<tbody
  data-testid="a-tbody"
/>

--------------------------------------------------
row:

Name "Cell 1 Cell 2 Cell 3":
<tr
  data-testid="a-row"
/>

--------------------------------------------------
cell:

Name "Cell 1":
<td
  data-testid="a-cell-1"
/>

Name "Cell 2":
<td
  data-testid="a-cell-2"
/>

Name "Cell 3":
<td
  data-testid="a-cell-3"
/>

--------------------------------------------------
form:

Name "a form":
<form
  aria-label="a form"
  data-testid="named-form"
/>

--------------------------------------------------
radio:

Name "":
<input
  data-testid="a-radio-1"
  type="radio"
/>

Name "":
<input
  data-testid="a-radio-2"
  type="radio"
/>

--------------------------------------------------
textbox:

Name "":
<input
  data-testid="a-input-1"
  type="text"
/>

Name "":
<input
  data-testid="a-input-2"
  type="text"
/>

Name "":
<textarea
  data-testid="a-textarea"
/>

--------------------------------------------------
`;
